crates/services/gas_price_service/src/v1/da_source_adapter/block_committer_costs.rs

```diff
@@ -6,57 +8,330 @@ use crate::v1::da_source_adapter::{
+    last_value: Option<RawDaBlockCosts>,
 }
 
 #[derive(Debug, Deserialize, Serialize, Clone, Default, PartialEq)]
 struct RawDaBlockCosts {
-    pub l2_block_range: core::ops::Range<u32>,
-    pub blob_size_bytes: u32,
-    pub blob_cost: u128,
+    // Sequence number (Monotonically increasing nonce)
+    pub sequence_number: u64,
+    // The range of blocks that the costs apply to
+    pub blocks_range: core::ops::Range<u64>,
+    // The DA block height of the last transaction for the range of blocks
+    pub da_block_height: DaBlockHeight,
+    // Rolling sum cost of posting blobs in wei
+    pub total_cost: u128,
+    // Rolling sum size of blobs (bytes)
+    pub total_size_bytes: u32,
 }
 
-impl From<RawDaBlockCosts> for DaBlockCosts {
-    fn from(raw: RawDaBlockCosts) -> Self {
-        DaBlockCosts {
-            l2_block_range: raw.l2_block_range,
-            blob_size_bytes: raw.blob_size_bytes,
-            blob_cost_wei: raw.blob_cost,
+impl<BlockCommitter> BlockCommitterDaBlockCosts<BlockCommitter> {
+    /// Create a new instance of the block committer da block costs source
+    pub fn new(client: BlockCommitter) -> Self {
+        Self {
+            client,
+            last_value: None,
         }
     }
 }
 
-impl BlockCommitterDaBlockCosts {
-    /// Create a new instance of the block committer da gas price source
+#[async_trait::async_trait]
+impl<BlockCommitter> DaBlockCostsSource for BlockCommitterDaBlockCosts<BlockCommitter>
+where
+    BlockCommitter: BlockCommitterApi,
+{
+    async fn request_da_block_cost(&mut self) -> DaBlockCostsResult<DaBlockCosts> {
+        let response;
+
+        if let Some(last_value) = &self.last_value {
+            response = self
+                .client
+                .get_costs_by_seqno(last_value.sequence_number + 1)
+                .await?;
+        } else {
+            // we have to error if we cannot find the first set of costs
+            response = self.client.get_latest_costs().await?;
+        }
+
+        if let Some(response) = response {
+            let res;
+            if let Some(last_value) = &self.last_value {
+                res = DaBlockCosts {
+                    l2_block_range: response.blocks_range.clone(),
+                    blob_size_bytes: response
+                        .total_size_bytes
+                        .checked_sub(last_value.total_size_bytes)
+                        .ok_or(anyhow!("Blob size bytes underflow"))?,
+                    blob_cost_wei: response
+                        .total_cost
+                        .checked_sub(last_value.total_cost)
+                        .ok_or(anyhow!("Blob cost wei underflow"))?,
+                };
+            } else {
+                res = DaBlockCosts {
+                    l2_block_range: response.blocks_range.clone(),
+                    blob_size_bytes: response.total_size_bytes,
+                    blob_cost_wei: response.total_cost,
+                };
+            }
```

Comment:
@MitchTurner if the block committer da block cost source is starting up for the first time, we will still get the `total_size_bytes` and `total_cost` from the block committer. is it reasonable to expect the consumers of this service to persist state and detect the first value that this source spits out? Or should we include the `last_state` in the `new` fn that constructs the `BlockCommitterDaBlockCosts`? I might lean to the latter approach since it is more testable, and allows consumers to actually bring their own persistence method.